extensions/Worker.Extensions.Http.AspNetCore/src/AspNetMiddleware/FunctionsEndpointDataSource.cs

```diff
@@ -0,0 +1,137 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text.Json;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Patterns;
+using Microsoft.Azure.Functions.Worker.Core.FunctionMetadata;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore.AspNetMiddleware
+{
+    internal class FunctionsEndpointDataSource : EndpointDataSource
+    {
+        private const string FunctionsApplicationDirectoryKey = "FUNCTIONS_APPLICATION_DIRECTORY";
+        private const string HostJsonFileName = "host.json";
+        private const string DefaultRoutePrefix = "api";
+
+        private readonly IFunctionMetadataProvider _functionMetadataProvider;
+        private readonly object _lock = new();
+
+        private static readonly JsonSerializerOptions _jsonSerializerOptions = new()
+        {
+            PropertyNameCaseInsensitive = true
+        };
+
+
+        private List<Endpoint>? _endpoints;
+
+        public FunctionsEndpointDataSource(IFunctionMetadataProvider functionMetadataProvider)
+        {
+            _functionMetadataProvider = functionMetadataProvider ?? throw new ArgumentNullException(nameof(functionMetadataProvider));
+        }
+
+        public override IReadOnlyList<Endpoint> Endpoints
+        {
+            get
+            {
+                if (_endpoints is null)
+                {
+                    lock (_lock)
+                    {
+                        _endpoints ??= BuildEndpoints();
+                    }
+                }
+
+                return _endpoints;
+            }
+        }
+
+        private List<Endpoint> BuildEndpoints()
+        {
+            List<Endpoint> endpoints = new List<Endpoint>();
+
+            string scriptRoot = Environment.GetEnvironmentVariable(FunctionsApplicationDirectoryKey) ??
+                           throw new InvalidOperationException("Cannot determine script root directory.");
+
+            var metadata = _functionMetadataProvider.GetFunctionMetadataAsync(scriptRoot).GetAwaiter().GetResult();
```

Comment:
I think the answer to this async-within-sync is maybe to re-evaluate how function metadata is accessed?

I am assuming function metadata is built once at startup. If that is true, then should there be a dedicated piece of startup that is responsible for building function metadata, then everything else accesses that already built metadata collection? Then you won't need to worry about async calls here.

This is a much larger change though, so for another PR, or even work item. In the meantime, I am not that concerned about the async call here as this is called just once.

Reply:
Yes -- that's the same thing I was thinking.

My plan is to follow-up after we get the next Preview out and see what we can do to make this API more re-usable. Because it involved changing the public surface in the worker, I didn't want to mess with that here.